### PR TITLE
adding player names to end-level feedback

### LINKED ISSUE #434: add '&' character between second-last and last player names in string

### PR DESCRIPTION
**Pull request by tongxiang**
#### What's this PR do?
A lot has been changed, but the basic functionality of this PR is simple. It: 
1. Adds a new feature: at the end of each level, all players receive a message telling them which players answered correctly (there's a flag right now to make sure this feature only appears for the story id of 301.)
2. Adds tests for this feature. 
3. Adds a new small module with string helper functions (obscenity checkers, etc.). When players enter their names, our apps now add a character limit (10 chars) and check for obscenities. Player names are also now added to user profiles on mobile commons with a more elegant syntax: "Tom, Barry, & Andrea" instead of "Tom, Barry, Andrea". 

#### Where should the reviewer start?
**Files to pay attention to:**
1. `playerNameHelpers.js` - this is where the `endLevelMessageWithSuccessPlayerNames.js` function is actually written, handling the meat of the end-level player name messaging logic. 
2. `SGcompetitiveStoryController.js`, `gameMessageHelpers.js`, `logicUserAction.js` - have been modified to support the implementation of this feature. 
3. `competitiveStoriesConfigModel.js` - modified to support the four new opt in paths required for the player name feature. 

**Tests files changed
1. `testHelperFunctions.js` now supports checking end-level messages
2. `GameAlphaStart.js` tests the validity of player name messages during the onboarding process. 
3. `BullyText2015ABTest.js` tests the validity of end-level player name messages. 
4. `_init.test.js` starts the new tests. 

**Files to pay less attention to:**
1. `/string-validators/index.js`, `DonorsChooseDonationController.js` - I moved the obscenity checkers out from the DonorsChoose library and into a new library.

#### How should this be manually tested?
`npm test`! I've manually tested this feature extensively with Postman creating a four-user game, but Alysha and @ArielSD will be QB-ing a manual testing strategy for this feature. 

#### What are the relevant tickets?
Closes #434, #433, #432, #430.

### CODE REVIEW

File: app/lib/sms-games/controllers/SGCompetitiveStoryController.js [hunk 1]
```diff
@@ -86,10 +90,17 @@ SGCompetitiveStoryController.prototype.createGame = function(request, response)
     return false;
   }
 
-  // Compile a new game document.
+  function truncName(name) {
+    if (!name) { return null; }
+    else if (stringValidator.containsNaughtyWords(name)) { return "IamNaughty"; }
```
Comment (jonuy) [inline app/lib/sms-games/controllers/SGCompetitiveStoryController.js]: I'd push this `"IamNaughty"` up into its own module-wide var like `NAME_CHAR_LIMIT`. Will make it easier to find if we want to change it.

File: app/lib/sms-games/controllers/playerNameHelpers.js [hunk 2]
```diff
@@ -115,4 +118,94 @@ function betaJoinNotifyAllPlayers(gameConfig, gameDoc, joiningBetaPhone) {
...
+  for (i = 0; i < allPlayers.length; i++) {
+    message.singleUserWithDelay(allPlayers[i].phone, endLevelMessage, delay, gameDoc._id, userModel, args);
+    emitter.emit('end-level-player-name-message', {'players_who_succeeded_at_end_level' : nameString, 'endLevelMessage': endLevelMessage});
+  }
+}
+
+// Removes last comma, add & to penultimate comma
+function removeLastCommaAddAnd(nameString) {
+  removeIndex = nameString.lastIndexOf(",");
+  return nameString.substring(0, removeIndex).replace(/, (?!.*?, )/, ' & ');
```
Comment (jonuy) [inline app/lib/sms-games/controllers/playerNameHelpers.js]: Hello, regex!

**Comment by jonuy** [comment]
I might have missed it when looking through this, what happens when this code runs on games that may not be setup to handle the personalized end-level messages? Does any special config need to happen?

Looking good :+1:

**Comment by tongxiang** [comment]
@jonuy the normal game runs for games not set up to handle end-level player name messages. (There's a flag set up to check that the game doc has a `story_id` of 301 before sending player name messages. Not the most ideal situation. I created [an issue](https://github.com/DoSomething/ds-mdata-responder/issues/438) to make sure I stay on top of removing this flag once we know we want this feature for all games.)

**Pull request merged**